### DISCUSSION TITLE
How to control the initial sample

### DISCUSSION THREAD

**Discussion by miaoL1** [netket/netket#1672] [Q&A] [answered]
My lattice model is composed of two disconnected graphs, such as bi-layer graphene：
```
L = 1 #number of unit cells
graph = nk.graph.DoubledGraph(nk.graph.Honeycomb([L,L],pbc=True))
N = graph.n_nodes #number of sites
hi = nkx.hilbert.SpinOrbitalFermions(N, s=1/2, n_fermions=N)
```
I want to keep electron number in each layer conserved and fixed at half-filling.
this sampler
```
nk.sampler.MetropolisExchange(hilbert=hi,graph=graph,d_max=1)
```
does help me keep electron number in each layer conversed, as the exchange only happens on points which are connected by edges in graph, and two layers are presented by disconnected graphs. However, the total charge difference, i.e. the electron number in each layer, is random and determined at the initial sample.

How can I guarantee that sampling only happens in subspace of half-filling in each layer?
I'd appreciate it if anyone can help.

**Comment by PhilipVinc** [comment]
I think it's documented [in the doc of SpinOrbitalFermions](https://netket.readthedocs.io/en/latest/api/_generated/hilbert/netket.experimental.hilbert.SpinOrbitalFermions.html). 

I believe you just need to declare `nkx.hilbert.SpinOrbitalFermions(N, s=1/2, n_fermions_per_spin=(N//2, N//2))`

**Reply by miaoL1** [reply]
I see, the parameter "n_fermions_per_spin" keeps the spin conserved. 
However, I want the sub-space which keeps the electron in two-layer half-filling.

**Reply by PhilipVinc** [reply, marked as answer]
```
hi = nkx.hilbert.SpinOrbitalFermions(N//2, s=1/2, n_fermions=N//2) * nkx.hilbert.SpinOrbitalFermions(N//2, s=1/2, n_fermions=N//2)
```

?

**Reply by miaoL1** [reply]
wow, I didn't expect to do this. Thanks!

**Reply by PhilipVinc** [reply]
Though this will change the encodings of the fermions, are the two spaces are concatenated. So you'll have 
`[layer 1 spin down..., layer 1 spin up..., layer 2 spin down..., layer 2 spin up...]` instead of the current one you have which is
`[layer 1 spin down, layer 2 spin down, layer 1 spin up, layer 2 spin up]`.

Also tensor Hilbert spaces don't probably play very well with Operators...

**Reply by PhilipVinc** [reply]
It might just be easier if you define a new `ExchangeRule` with a custom overriden `init_state` that does what you want.

**Reply by miaoL1** [reply]
by the way, I'm wondering what should I do if I both want the charge conservation in each two layer, and overall spin conservation(such as total spin = 0).
